source/fuzz/fuzzer_util.cpp

```diff
@@ -205,6 +205,123 @@ opt::BasicBlock::iterator GetIteratorForBaseInstructionAndOffset(
   return nullptr;
 }
 
+// Helper method to figure out whether there is a back-edge-free path from
+// |source| to |dest|.
+bool CanReachWithoutTraversingLoopBackEdge(opt::IRContext* context,
+                                           opt::BasicBlock* source,
+                                           opt::BasicBlock* dest) {
+  // First, check that the destination block occurs after the source block in
```

Comment:
This isn't true

```
for (;;) {
  // a
  if (cond) {
    // b
  } else {
    // c
  }
  // d
}
// e
```
Blocks `b` and `c` may appear after the loop merge `e` in the module.  So `b` and `c` could reach `d` and `e` without traversing a back-edge, but not be found in the iteration.

There is an optimizer utility to order the blocks in "structured order" so that all the blocks contained in a construct are ordered before the merge.

There is a similar issue in the iteration in the other function, but I think it will generally result in overly conservative decisions.

Reply:
Thanks - you're right of course!

I have done a commit that adds a test exposing the problem you have identified, and that includes a new algorithm, designed in collaboration with @paulthomson, which we believe does the right thing.

It is inspired by the structured order algorithms in the optimizer - in particular the idea of doing a depth-first search where merge blocks are visited before successors.

I personally don't think it's worth trying to directly re-use code here.